grype/pkg/provider.go

```diff
@@ -28,29 +29,43 @@ func Provide(userInput string, scopeOpt source.Scope) ([]Package, Context, error
 		syftProvider, // important: we should try syft last
 	}
 
-	// aggregate stdin into a buffer that can be used across multiple providers dynamically
-	var previousStdin bytes.Buffer
+	// capture stdin bytes, so they can be used across multiple providers
+	capturedStdin := bytesFromStdin()
 
-	for _, p := range providers {
-		cfg := providerConfig{
-			userInput: userInput,
-			scopeOpt:  &scopeOpt,
-		}
-
-		if internal.IsPipedInput() && userInput == "" {
-			// this is a hint to all providers that there is already a reader available, don't try and derive one from user input
-			// however, the user input may be useful in situations where the reader isn't provided
-
-			// this reader is a combination of previous bytes read from stdin by other providers as well as what still is
-			// available from stdin. The Tee reader is to ensure that any read bytes from stdin are preserved.
-			cfg.reader = io.MultiReader(&previousStdin, io.TeeReader(os.Stdin, &previousStdin))
```

Comment:
The main goal here was to simplify. To be honest, I was having trouble keeping track of the logic, which was notable to me because the lack of availability of SBOM JSON readers was central to these bugs. For now, we only have one provider that makes use of the stdin bytes reader, so the `multireader` + `teereader` weren't being leveraged beyond providing an `io.Reader` for stdin. I could see this changing if we see a need for more providers, and at that point, it'd be good to reassess the symmetry we're seeing among the providers.

If there's a more straightforward way to improve this code, I'd welcome that!